### PR TITLE
Refactor cli

### LINKED ISSUE #117: Create CLI in cli.py and in each script
At the moment we have the CLI defined in `__main__.py` to allow for the following use `python -m grants_tagger.train` which I personally never use but regardless we can define it more cleanly in its own file `cli.py` and then import the `app` in `__main__.py`.

Additionally each script should have its own cli that someone can use for example `python grants_tagger/train.py` and that CLI should be imported inside cli and assembled into the main cli we expose.

### PR DESCRIPTION
**Pull request by pdan93**
## Description
Move CLI logic from __main__.py to cli.py. Have every script declare it's own cli, and the main cli.py just imports those for the complete CLI

Fixes #117 

## Checklist

- [x] Linked to Notion or GitHub issue
- [x] Added tests
- [ ] Updated README
- [ ] DVC up to date

### CODE REVIEW

File: grants_tagger/cli.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+from grants_tagger.tune_threshold import tune_threshold_cli
+from grants_tagger.optimise_params import tune_params_cli
+from grants_tagger.download_epmc import download_epmc_cli
+from grants_tagger.download_model import download_model_cli
+from grants_tagger.explain import explain_cli
+
+app = typer.Typer(add_completion=False)
+
+
+# Move to train and import from there
```
Comment (nsorros) [inline grants_tagger/cli.py:39]: we can remove that now

**Comment by nsorros** [comment]
⚠️ Don't forget to include the app in `__main__` so `python -m grants_tagger` works

File: tests/test_cli.py [hunk 2]
```diff
@@ -133,19 +134,16 @@ def test_train_command():
         result = runner.invoke(app, [
             "train",
             data_path,
-            model_path,
             label_binarizer_path,
+            model_path,
```
Comment (nsorros) [inline tests/test_cli.py:138]: did we change the order? this might affect `dvc.yaml`

Comment (pdan93) [reply tests/test_cli.py:138]: resolving :)

File: grants_tagger/train_with_sagemaker.py [hunk 3]
```diff
@@ -141,3 +148,100 @@ def train_with_sagemaker(instance_type="local", config_version=None, **kwargs):
...
+        threshold=threshold,
+        data_format=data_format,
+        sparse_labels=sparse_labels,
+        cache_path=cache_path,
+        instance_type=instance_type,
+        config_version=config_version,
+    )
+
+    duration = time.time() - start
+    print(f"Took {duration:.2f} to train")
```
Comment (nsorros) [inline grants_tagger/train_with_sagemaker.py:238]: This was supposed to measure how much time train took. It should go back to the main train in cli.py

**Review by nsorros — CHANGES_REQUESTED** [review]
There are two things that need attention

- ensuring `dvc.yaml` works
- the training time info should go back to main train

Other than that, its looking good and its ready to be merged

File: tests/test_cli.py (hunk 2 above)
Comment (pdan93) [reply tests/test_cli.py:138]: you know what the weird thing is, dvc actually had the correct order, just the test had the wrong one. Meaning it was passing until now with the wrong order somehow

Comment (pdan93) [reply tests/test_cli.py:138]: @nsorros

File: grants_tagger/train_with_sagemaker.py (hunk 3 above)
Comment (pdan93) [reply grants_tagger/train_with_sagemaker.py:238]: Reason I moved it in both train scripts is so it measures it also when you call the individual scripts, doesn't it make more sense like that?

**Comment by pdan93** [comment]
> warning Don't forget to include the app in `__main__` so `python -m grants_tagger` works

already included

**Comment by pdan93** [comment]
With last commit I verified order of all arguments in dvc.yaml and there was only one change in the `preprocess_wellcome_science` step. Everything else seems in order, but I have no way of running `dvc repro`, you'll have to do it @nsorros

File: grants_tagger/train_with_sagemaker.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger/train_with_sagemaker.py:238]: Actually it mostly makes sense for train that happens locally so remove it from sagemaker and only have it in train_cli. Also in train_cli you do not pass the train_info. We want to save this information in a file and you are only printing in screen

File: tests/test_cli.py [hunk 4]
```diff
@@ -133,19 +134,16 @@ def test_train_command():
...
             "--approach",
             "tfidf-svm",
             "--parameters",
             "{'tfidf__min_df': 1, 'tfidf__stop_words': None}",
-            "--train-info",
```
Comment (nsorros) [inline tests/test_cli.py:142]: part of the reason you are not catching the problem i mentioned above is you remove that test. btw there is a problem with that test see #154. could we also fix that as part of this PR

File: tests/test_cli.py (hunk 2 above)
Comment (pdan93) [reply tests/test_cli.py:138]: I don't get it, it shouldn't have based on what's in the train code.

**Review by nsorros — APPROVED** [review]
This looks ready but let me run dvc repro overnight to iron any issues that we did not find in the code review. Also remove the time info from sagemaker.

**Comment by pdan93** [comment]
> om sagemaker. And resolve the conflict

done and done :)

**Pull request merged**